client/src/schema/query.ts

```diff
@@ -10,43 +11,121 @@ export type QueryOptions<T> = {
   sort?: SortFilter<T> | SortFilter<T>[];
 };
 
-type QueryOrConstraint<T extends XataRecord, R extends XataRecord> = Query<T, R> | Constraint<T>;
+export type FilterExpression = {
+  $exists?: string;
+  $existsNot?: string;
+  $any?: FilterList;
+  $all?: FilterList;
+  $none?: FilterList;
+  $not?: FilterList;
+} & {
+  [key: string]: FilterColumn;
+};
+
+export type FilterList = FilterExpression | FilterExpression[];
+export type FilterColumn = FilterColumnIncludes | FilterPredicate | FilterList;
+
+/**
+ * @maxProperties 1
+ * @minProperties 1
```

Comment:
Interesing! Didn't know this existed. Is this used by TypeScript and/or vscode?

Reply:
This is generated from the openapi spec, unfortunately it's not used in TS/VSCode.

I've done things like these with helpers before in TS (maybe as an improvement for the future):

```ts
export type NonEmptyArray<T> = T[] & { 0: T };

export type RequireAtLeastOne<T, Keys extends keyof T = keyof T> = Pick<T, Exclude<keyof T, Keys>> &
    {
        [K in Keys]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<Keys, K>>>;
    }[Keys];
```